service/matching/matcher.go

```diff
@@ -115,25 +126,111 @@ func (tm *TaskMatcher) Offer(ctx context.Context, task *internalTask) (bool, err
 	}
 }
 
+// OfferQuery offers a query task to a potential consumer (poller). If the task
+// is successfully matched, this method will return the query response. Otherwise
+// it returns error
+func (tm *TaskMatcher) OfferQuery(ctx context.Context, task *internalTask) ([]byte, error) {
+	select {
+	case tm.queryTaskC <- task:
+		<-task.responseC
+		return nil, nil
+	default:
+	}
+
+	fwdrTokenC := tm.fwdrAddReqTokenC()
+
+	for {
+		select {
+		case tm.queryTaskC <- task:
+			<-task.responseC
+			return nil, nil
+		case token := <-fwdrTokenC:
+			resp, err := tm.fwdr.ForwardQueryTask(ctx, task)
+			token.release()
+			if err == nil {
+				return resp.QueryResult, nil
+			}
+			if err == errForwarderSlowDown {
+				// if we are rate limited, try only local match for the
+				// remainder of the context timeout left
+				fwdrTokenC = noopForwarderTokenC
+				continue
+			}
+			return nil, err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
 // MustOffer blocks until a consumer is found to handle this task
 // Returns error only when context is canceled or the ratelimit is set to zero (allow nothing)
 // The passed in context MUST NOT have a deadline associated with it
 func (tm *TaskMatcher) MustOffer(ctx context.Context, task *internalTask) error {
 	if _, err := tm.ratelimit(ctx); err != nil {
 		return err
 	}
+
+	// attempt a match with local poller first. When that
+	// doesn't succeed, try both local match and remote match
 	select {
 	case tm.taskC <- task:
 		return nil
 	case <-ctx.Done():
 		return ctx.Err()
+	default:
+	}
+
+forLoop:
+	for {
+		select {
+		case tm.taskC <- task:
+			return nil
+		case token := <-tm.fwdrAddReqTokenC():
+			childCtx, cancel := context.WithDeadline(ctx, time.Now().Add(time.Second*2))
+			err := tm.fwdr.ForwardTask(childCtx, task)
+			cancel()
+			token.release()
+			if err != nil {
+				// forwarder returns error only when the call is rate limited. To
+				// avoid a busy loop on such rate limiting events, we only attempt to make
+				// the next forwarded call after this childCtx expires. Till then, we block
+				// hoping for a local poller match
+				select {
+				case tm.taskC <- task:
+					return nil
+				case <-childCtx.Done():
```

Comment:
If I am not mistaken the cancel you call above will cause the child context to be done here, so this wont block for local poller, it will exit the select immediately.

Reply:
good match. Missed this as I moved around code.